Add monitoring for how often nodes get killed

jguipi opened the pull request:
How to categorize this PR?
/area monitoring
/kind enhancement

What this PR does / why we need it:
- Add a new rule to monitor nodes that are in the NotReady state for more than 10 minutes due to the kubelet not responding
- Add a new dashboard in the "Node/Worker Pool Overview" page to monitor node in the NotReady state trend. 

Which issue(s) this PR fixes:

Special notes for your reviewer:

This metric is intended to compare the average lifespan of nodes within existing clusters, with the aim of detecting potential issues that may not be readily apparent through the self-healing process that occurs when nodes become "NotReady". The metric will also enable identification of any changes in trends, such as an upward shift in the event of a new Gardenlinux version release, which may assist in more rapid root cause analysis.

Release note:

```other operator
- Add a new rule to monitor nodes that are in the NotReady state for more than 10 minutes due to the kubelet not responding
- Add a new dashboard in the "Node/Worker Pool Overview" page to monitor node in the NotReady state trend. 
```

jguipi committed:
Add graph to display Nodes that have been in the NotReady state for over 10 minutes.
>
>
Co-authored-by: dmahmalat <dany.mahmalat@sap.com>

jguipi committed:
Add the NodeKubeletDown alert.
>
>
Co-authored-by: dmahmalat <dany.mahmalat@sap.com>

rfranzke commented:
/cc @istvanballok @rickardsjp

acumino commented on charts/seed-monitoring/charts/core/charts/prometheus/rules/node-exporter.rules.yaml:
```diff
@@ -1,6 +1,17 @@
 - name: node-exporter.rules
   rules:
+  - alert: NodeKubeletDown
+    expr: count_over_time((sum by (node)(kube_node_status_condition{condition="Ready",status!="true"}) == 1)[10m:]) == 9
```
Had a combined look with @ScheererJ and @shafeeqes. We have following concerns:
- Query for Prometheus and Grafana is not the same.
- Does the node get deleted/removed if it is in `NotReady` Condition for 10 min, In that case we should reduce the timing in query to 9 or 8 minutes.
- Should the dashboard be part of the `node-exporter`,  as of now it comes as a separate block that is not properly aligned.
- Node count scale can't have a decimal value.
<img width="1651" alt="Screenshot 2023-03-21 at 3 52 35 PM" src="https://user-images.githubusercontent.com/28653770/226578269-8035a6dd-b433-4668-9cdf-8e0fa3ded0a8.png">

/cc @istvanballok @rickardsjp

jguipi replied:
To address some of the aforementioned concerns, we have conducted several rounds of testing. The reason behind setting the threshold at 10 minutes is to minimize the occurrence of false positives. Our testing has shown that a variety of factors can trigger a node to be deemed "not ready," including updates to k8s version, GardenLinux, and patch versions.

In our analysis, we have considered a worst-case scenario, assuming an empty cluster. In such a case, any event lasting longer than 10 minutes may be deemed meaningful, particularly if it persists for an extended period. Our objective is to filter out non-essential events and only flag nodes that are potentially experiencing genuine issues.

Not sure if that dashboard should be elsewhere. I'm open to the discussion. 

The node count scale will adapt when values are generated. 

---


The explanation of the expression : 

Breaking down the PromQL expression:
The expression is:
expr: count_over_time((sum by (node)(kube_node_status_condition{condition="Ready",status!="true"}) == 1)[10m:]) == 10

kube_node_status_condition{condition="Ready",status!="true"} == 1: prints a table of nodes where the status is not true (printed are nodes with status false or unknown) (== 1 means that the lines on the table satisfy this expression)

((kube_node_status_condition{condition="Ready",status!="true"}) == 1)[10m:]: For EACH data point, performs the above query every evaluation interval for the past 10 minute window. Here, we have evaluation_interval: 1m, so this would check once every minute for the past 10 minutes. Because == 1 exists, this will only show lines where this expression was true. So if we have 10 elements, the condition was true for 10 minutes.

(sum by (node)(kube_node_status_condition{condition="Ready",status!="true"}) == 1)[10m:]: (sum by (node) groups the results by node.

count_over_time((sum by (node)(kube_node_status_condition{condition="Ready",status!="true"}) == 1)[10m:]) == 10: count_over_time counts how many times the condition is true (instead of a 1 at every evaluation interval, this counts how many times the query returns 1). The == 10 part only considers nodes that have had this condition true for every minute in the past 10 minute evaluation time.

ScheererJ replied:
Have you actually seen nodes being not ready for longer than 10 minutes (with the default settings in place)? I wonder if we can monitor this as a node might automatically be replaced after 10 minutes (using default settings). As the sampling rate of our metrics is only once per minute, a node might hit the 10 minute not-ready mark well before the tenth metric is being scraped. If the node deletion then happens in a not so graceful manner we might not even get the tenth metric.

Therefore, I would propose to reduce it to 8. It might show some nodes that will not be replaced, but it will more likely show the nodes that get replaced.

> Not sure if that dashboard should be elsewhere. I'm open to the discussion.

It can be in the current dashboard, but it should be aligned properly. Currently it is only filling half the screen with the left side being empty.

> The node count scale will adapt when values are generated.

The node count should be an integer value. There will be no `.5` nodes non-ready.

ScheererJ replied:
Please also note that the amount of time until a non-ready node gets replaced is configurable. Therefore, a static check like this might always have false positives if a cluster owner choses a different value, e.g. 20 minutes.

jguipi replied:
Yes, it actually happened in the past if I recall correctly, at some point there was an issue with a new Gardenlinux version, and we needed to roll back. 

Seeing a lot of restart could have helped us identify that faster. I think the goal here is more to establish a trend of how many nodes get notReady on a regular basis and see if there's a big trend change at some point that could help us identify a recent change that has been made on our side.

I agree that there's no one size to fit all regarding the configurable parameter. Maybe @BeckerMax can share a little more on that topic here.

jguipi committed:
Update node-pool-dashboard.json

Remove decimal number

jguipi commented:
Moved the dashboard a bit higher and removed the decimal number. Will discuss this Pr with the monitoring team on Friday. 
<img width="1618" alt="image" src="https://user-images.githubusercontent.com/11049349/228620656-8f069394-4fa7-4316-bf07-f62fc24508b4.png">

jguipi committed:
Remove decimal number in the Nodes not ready dashboard and change dashboard location

istvanballok reviewed (commented):
The insight by @ScheererJ that the machine health timeout is [configurable](https://github.com/gardener/machine-controller-manager/blob/64ed53ef4dbd48be89572d671a3d6523bd765600/pkg/util/provider/app/options/options.go#L100) renders this approach that @rickardsjp and I suggested to @jguipi and Dany which is based on empirical values and kube-state-metrics metrics about the node status not so promising, because we can not clearly identify the scenario when a node fails and is recreated by the machine-controller due to being unhealthy for too long compared to scenarios when a node is deleted due to cluster scaling or kubernetes or OS upgrades.
Instead, we should try to let the machine controller expose a new metric like

```
HELP mcm_machines_failed_total A counter for failed machines that have been recreated by mcm
```

Would that be a feasible addition to the machine controller, @himanshu-kun, possibly as a contribution by @jguipi?

With @rickardsjp

himanshu-kun commented:
> Would that be a feasible addition to the machine controller, @himanshu-kun, possibly as a contribution by @jguipi?

Yes please feel free to do so. I think its a valuable addition. We already have a `mcm_machine_deployment_failed_machines` and `mcm_machine_set_failed_machines` metrics but they are for "failed last operation machines" and not for "failed phase machines" . The name has to be updated.
Just mentioning so that there is no confusion.

jguipi commented:
Will close this PR. We will try to implement this by adding a new metric to the MCM to be more accurate and take the `MachineHealthTimeout` into consideration.
